<a href="https://colab.research.google.com/github/valeriarpo/Business-Analytics/blob/main/2_RETO1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd

from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix


In [ ]:
nxl = ""
XDB = pd.read_excel(nxl, sheet_name= 1)
XDB        #6705 datos
XDB=XDB.dropna()     #Quita celdas vacias

XD=XDB[["Weekday","ViolCat","ClearWeather","Month","Daylight"]]
XD

yd=XDB[["Highway"]]     #Modelo identifique si accidente sucedio en autopista o no
yd

conteo = yd.value_counts()
porcentaje_categorias = (conteo/len(XD))*100
print(porcentaje_categorias)                               #Porcentaje de accidentes en autopista 28%  #Porcentaje accidentes en NO autopista/zonas rurales 72%





Highway
0          72.856078
1          27.143922
Name: count, dtype: float64


In [ ]:
mnb= GaussianNB()
mnb.fit(XD,yd)
u = mnb.theta_      #son las medias
print("las medias de las variables son:\n", u)
sigma=np.sqrt(mnb.var_)             #desviación
print("las valores de la desviación son:\n", sigma)


Limite_Superior= u+sigma
print("los limites superiores son:\n", Limite_Superior)
Limite_Inferior= u-sigma
print("los limites inferiores son:\n", Limite_Inferior)


ydp=mnb.predict(XD)


cm=confusion_matrix(yd,ydp)
print("La matriz de confusión es::\n", cm)

VN=cm[0,0]
FP=cm[0,1]
FN=cm[1,0]
VP=cm[1,1]
print("VN=",VN,"FP=",FP,"FN=",FN,"VP=",VP)

Exactitud=(VP+VN)/len(XD)
print("la exactitud es=",Exactitud)

Tasa_de_error= (FP+FN)/len(XD)
print("La tasa de error del modelo es: ", Tasa_de_error)

sensibilidad= VP/(VP+FN)
print("La sensibilidad del modelo es: ", sensibilidad)

especificidad= VN/(VN+FN)
print("La especificidad del modelo es: ", especificidad)

pres= VP/(VP+FP)
print("La precisión del modelo es: ", pres)

PredNeg= VN/(VN+FN)
print("La predicción negativa del modelo es: ", PredNeg)

XSC = [[7,1,1,3,0]]
ydsc=mnb.predict(XSC)
print("La predicción del modelo es: ", ydsc)


las medias de las variables son:
 [[3.88945752 6.7778915  0.85261003 5.52548618 0.6814739 ]
 [3.82417582 4.31538462 0.79450549 5.47417582 0.70934066]]
las valores de la desviación son:
 [[1.92831125 3.63475806 0.35449426 3.27379337 0.46590475]
 [1.82800803 2.36793707 0.40406253 3.26215279 0.45406663]]
los limites superiores son:
 [[ 5.81776877 10.41264957  1.20710429  8.79927955  1.14737865]
 [ 5.65218385  6.68332168  1.19856802  8.73632861  1.16340729]]
los limites inferiores son:
 [[1.96114628 3.14313344 0.49811577 2.25169281 0.21556915]
 [1.9961678  1.94744755 0.39044297 2.21202304 0.25527403]]
La matriz de confusión es::
 [[4560  325]
 [1563  257]]
VN= 4560 FP= 325 FN= 1563 VP= 257
la exactitud es= 0.7184190902311708
La tasa de error del modelo es:  0.2815809097688292
La sensibilidad del modelo es:  0.1412087912087912
La especificidad del modelo es:  0.7447329740323371
La precisión del modelo es:  0.4415807560137457
La predicción negativa del modelo es:  0.7447329740323371
La predi

/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


In [ ]:
XDBn=pd.read_excel(nxl,sheet_name=0)
XDBn.head(100)

XDn=XDBn[["Weekday","ViolCat","ClearWeather","Month","Daylight"]]


ydn=XDBn[["Highway"]]

ydpn=mnb.predict(XDn)
ydpn

XDsal=np.column_stack((XDn,ydpn,ydn))
df_sal=pd.DataFrame(XDsal)
df_sal.to_excel("Informe.xlsx")


# **Analísis de Resultado**
Los resultados obtenidos por el modelo propuesto para la predicción de accidentes automovilísticos, en autopista o no, en el condado de San Diego son los siguientes: el modelo analizó un total de 6705 datos, en los cuales mostró una exactitud del 72%. Este es un porcentaje lo suficientemente alto para considerar la información analizada como verídica; por lo tanto, la tasa de error del modelo es del 28%
Esto se refleja en el análisis, donde, del total de datos, el modelo acertó en la predicción de 4560 accidentes en zonas rurales y se equivocó 325 veces en el análisis de accidentes que no ocurrieron en la autopista (72% del total de datos). Asimismo, el modelo acertó en la predicción de 1,563 accidentes en la autopista y se equivocó 257 veces (28% del total de datos).
En cuanto al intervalo de días de la semana en que ocurren los accidentes, este varía desde el día 1.9 (que se puede aproximar al día 2) hasta el día 5. Tomando en cuenta la media, se podría decir que los accidentes suelen suceder a mitad de semana. Además, en el análisis de violcat, el intervalo de accidentes en zonas rurales es de 3.1 a 10.4, mientras que el intervalo para accidentes en autopista es de 1.9 a 6.6.
El modelo también cuenta con una sensibilidad del 14%, lo cual significa que 44% de los accidentes positivos han sido predichos correctamente como positivos. TIene specificidad del 93%, lo que indica los accidentes negativos que han sido predichos correctamente como negativos. La precisión del modelo al predecir los resultados es del 44%.
Finalmente, para un sujeto con las características analizadas anteriormente, su predicción de accidente en una autopista es negativa, lo cual nos lleva a concluir que vive en una zona rural.

Adicionalmente se concluye que 4885 de los datos corresponden a accidentes rurales y 1820 a accidentes en autopista.

